plugins/catalog-backend/migrations/20210408154713_add_attachments.js

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// @ts-check
+
+/**
+ * @param {import('knex').Knex} knex
+ */
+exports.up = async function up(knex) {
+  await knex.schema.createTable('entities_attachments', table => {
+    table.comment('Binary attachments to a entity');
+    table
+      .text('originating_entity_id')
+      .references('id')
+      .inTable('entities')
+      .onDelete('CASCADE')
+      .notNullable()
+      .index('originating_entity_id_idx')
+      .comment('The uid of the related entity');
+    table
+      .text('key')
+      .notNullable()
+      .comment('The name of the attachment, unique for a single entity');
+    table
+      .binary('data')
+      .notNullable()
+      .comment('The binary data of the attachment');
+    table
+      .text('content_type')
+      .notNullable()
+      .comment('The content type of the attachment for serving over HTTP');
+    // TODO: Store etag?
```

Comment:
👍 Also had some discussions with @freben around these optimization. For now I would skip this. But the database part already has the "keep this attachment around" feature, we just have to wire it up into the processor ones it makes sense.

Reply:
Btw, just noticed it. The topic etag is used in two places in the PR. On the one side the etags should be used to avoid repeated ingestion and speed up the process. The other etag (this one here) is responsible for serving the content on the Backstage api (like we already have for entities). The later is still something I plan to include in this PR.

Reply:
So we could reuse the "incoming" (remote) etag as our own API-presented etag, or calculate one ourselves as the hash of the data if there wasn't a remote one right?

When doing the etag check on incoming API requests, can the database query perhaps include something like `SELECT (IF etag = ? THEN NULL ELSE data) AS dataOrNull, etag ... WHERE key = ?` (pseudocode) so that we can discern between 404 conditions (no rows returned) and etag matches (get a null to avoid paying for the blob transfer from the db to the service)

Reply:
> So we could reuse the...
I would not reuse the etag if the part that contains the state is separate. Instead I would generate one on our own. Hash would be ideal, especially as long as we don't have the caching in the ingestion.

> When doing the etag check...

I'm just implementing that and choose just to do a query only for the metadata before that. So I can implement the caching logic in the router and keep it out of the database stuff. But I can also try to more complex query, I'm curious how easy it is to implement this with knex.

Reply:
There's always `raw` with knex. Just gotta have tests to see that it works in sqlite as well as postgres. It would be good to avoid two round trips to the database, but we can always start simple and iterate